sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/WALUtils.scala

```diff
@@ -0,0 +1,286 @@
+    val compressed = new LZ4CompressionCodec(sparkConf).compressedInputStream(inputStream)
+    new DataInputStream(compressed)
+  }
+
+
+  def writeUpdateToDeltaFile(
+                                      output: DataOutputStream,
+                                      key: UnsafeRow,
+                                      value: UnsafeRow): Unit = {
+    val keyBytes = key.getBytes()
+    val valueBytes = value.getBytes()
+    output.writeInt(keyBytes.size)
+    output.write(keyBytes)
+    output.writeInt(valueBytes.size)
+    output.write(valueBytes)
+  }
+
+  def writeRemoveToDeltaFile(output: DataOutputStream, key: UnsafeRow): Unit = {
+    val keyBytes = key.getBytes()
+    output.writeInt(keyBytes.size)
+    output.write(keyBytes)
+    output.writeInt(-1)
+  }
+
+  def finalizeDeltaFile(output: DataOutputStream): Unit = {
+    output.writeInt(-1)  // Write this magic number to signify end of file
+    output.close()
+  }
+
+  def updateFromDeltaFile( fm: CheckpointFileManager,
+                           fileToRead: Path,
+                           keySchema: StructType,
+                           valueSchema: StructType,
+                           newRocksDb: OptimisticTransactionDbInstance,
+                           sparkConf: SparkConf): Unit = {
+    var input: DataInputStream = null
+    val sourceStream = try {
+      fm.open(fileToRead)
+    } catch {
+      case f: FileNotFoundException =>
+        throw new IllegalStateException(
+          s"Error reading delta file $fileToRead of $this: $fileToRead does not exist", f)
+    }
+    try {
+      input = decompressStream(sourceStream, sparkConf)
+      var eof = false
+
+      while(!eof) {
+        val keySize = input.readInt()
+        if (keySize == -1) {
+          eof = true
+        } else if (keySize < 0) {
+          newRocksDb.abort
+          newRocksDb.close()
+          throw new IOException(
+            s"Error reading delta file $fileToRead of $this: key size cannot be $keySize")
+        } else {
+          val keyRowBuffer = new Array[Byte](keySize)
+          ByteStreams.readFully(input, keyRowBuffer, 0, keySize)
+
+          val keyRow = new UnsafeRow(keySchema.fields.length)
+          keyRow.pointTo(keyRowBuffer, keySize)
+
+          val valueSize = input.readInt()
+          if (valueSize < 0) {
+            newRocksDb.remove(key = keyRow)
+          } else {
+            val valueRowBuffer = new Array[Byte](valueSize)
+            ByteStreams.readFully(input, valueRowBuffer, 0, valueSize)
+            val valueRow = new UnsafeRow(valueSchema.fields.length)
+            // If valueSize in existing file is not multiple of 8, floor it to multiple of 8.
+            // This is a workaround for the following:
+            // Prior to Spark 2.3 mistakenly append 4 bytes to the value row in
+            // `RowBasedKeyValueBatch`, which gets persisted into the checkpoint data
```

Comment:
is there a JIRA on this?

Reply:
https://issues.apache.org/jira/browse/SPARK-21271

I have abstracted out utils related to WAL from the class [HDFSBackedStateStoreProvider](https://github.com/apache/spark/blob/f0f2f8d3ce8b0a4aa62134090ba90bcf39d97dc8/sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/HDFSBackedStateStoreProvider.scala#L549) so that it can be used elsewhere.